### PR TITLE
Mech ABI Prelim: Unify Mechanism Interfaces

### PR DESCRIPTION
**Pull request by thorstenhater**
As a stepping stone for the ABI feature we start by unification of the `public` and
`protected` interfaces of the mechanism classes. These interfaces will be merged
into the public API ('public' as in: exposed via `arbor/include/arbor`). 

This works towards #1335 and extends #1179. 

## Interface Methods

`nrn_` have been renamed
- `nrn_init`: `init` (would have preferred initialize, but that is taken)
- `deliver_events(event_stream)`: `apply_events` (to avoid a name clash)
- `nrn_state`: `advance_state`
- `nrn_current`: `compute_currents`

## Status
- [X] Coherent interface
  - [X] Adjust modcc
  - [X] CPU tests pass
  - [X] CPU SIMD tests pass
  - [X] GPU tests pass
- [X] Rename `nrn_` methods
- [x] Clean-up
  - [x] Remove typedefs for index_type etc
  - [X] Shuffle `ion_state_view` to public interface
  - [x] Migrate `copy_extend` to a proper place out of `multicore/mechanism.hpp`. Maybe `memory` or `range_util`.
  - [X] remove `vec_t_to_`

## To be addressed later
 The multicore backends might want to use the `ppack` construction, too, which further unifies the backend architectures. However, this change might be much more controversial, so I am deferring this to another PR.

### CODE REVIEW

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

File: arbor/backends/event_delivery.md [hunk 1]
```diff
@@ -114,7 +114,7 @@ For `fvm_multicell` one integration step comprises:
 Towards the end of the integration period, an integration step may have a zero _dt_
 for one or more cells within the group, and this needs to be handled correctly:
 
-*   Generated mechanism `nrn_state()` methods should be numerically correct with
+*   Generated mechanism `advance_state()` methods should be numerically correct with
```
Comment (bcumming) [inline arbor/backends/event_delivery.md:117]: The new names (`advance_state`, `compute_currents`, `init`, `apply_events`) are find with me.

Though I would be happy to stick with `deliver_events`.

File: arbor/backends/gpu/mechanism_ppack_base.hpp [hunk 2]
```diff
@@ -3,30 +3,17 @@
...
-    value_type* ionic_charge;
-};
-
 // Parameter pack base:
-
 struct mechanism_ppack_base {
     using value_type = fvm_value_type;
     using index_type = fvm_index_type;
-    using ion_state_view = ::arb::gpu::ion_state_view;
+    using ion_state_view = ::arb::ion_state_view;
```
Comment (bcumming) [inline arbor/backends/gpu/mechanism_ppack_base.hpp:16]: Is `arb::ion_state_view` is not common between all mechanism descriptions? If so, then we don't need this using statement at all.

File: arbor/backends/multicore/mechanism.cpp [hunk 3]
```diff
@@ -163,22 +163,33 @@ void mechanism::instantiate(unsigned id, backend::shared_state& shared, const me
...
-        auto ion_binding = value_by_key(overrides.ion_rebind, i.first).value_or(i.first);
-
-        ion_state* oion = ptr_by_key(shared.ion_data, ion_binding);
-        if (!oion) {
-            throw arbor_internal_error("multicore/mechanism: mechanism holds ion with no corresponding shared state");
-        }
+    {
+        auto table = ion_index_table();
+        // Allocate bulk storage
+        auto count = (table.size() + 1)*width_padded_;
```
Comment (bcumming) [inline arbor/backends/multicore/mechanism.cpp:169]: Why `table.size()+1`? Shouldn't it be enough to allocate enough space for `table.size()`? In the loop below `table.size()` indexes are created, not `table.size()+1`.

File: arbor/include/arbor/mechanism.hpp [hunk 4]
```diff
@@ -49,8 +53,11 @@ class mechanism {
     // Non-global parameters can be set post-instantiation:
     virtual void set_parameter(const std::string& key, const std::vector<fvm_value_type>& values) = 0;
 
+    // Peek into state variable
+    virtual fvm_value_type* field_data(const std::string& var) = 0;
```
Comment (bcumming) [inline arbor/include/arbor/mechanism.hpp:57]: the `value_type` is typedefed above, but we are using the `fvm_value_type` here. There should be consistency. Maybe we don't need the typedefs?

File: modcc/mechanism.hpp [hunk 5]
```diff
@@ -6,7 +6,7 @@
     Abstract base class for all mechanisms
 
     This works well for the standard interface that is exported by all mechanisms,
-    i.e. nrn_jacobian(), nrn_current(), etc. The overhead of using virtual dispatch
+    i.e. nrn_jacobian(), compute_currents(), etc. The overhead of using virtual dispatch
```
Comment (bcumming) [inline modcc/mechanism.hpp:9]: oops, we can remove `nrn_jacobian` from this comment!

**Review by bcumming — CHANGES_REQUESTED** [review]
Looks good at a high level.

There are just a few small fixes/suggestions.

File: arbor/backends/gpu/mechanism_ppack_base.hpp (hunk 2 above)
Comment (thorstenhater) [reply arbor/backends/gpu/mechanism_ppack_base.hpp:16]: Yes, that was a mechanical change that is probably redundant.

File: arbor/backends/multicore/mechanism.cpp (hunk 3 above)
Comment (thorstenhater) [reply arbor/backends/multicore/mechanism.cpp:169]: Yes.

File: modcc/mechanism.hpp (hunk 5 above)
Comment (thorstenhater) [reply modcc/mechanism.hpp:9]: Yes, it's gone locally, as discussed on Slack.

File: arbor/include/arbor/mechanism.hpp (hunk 4 above)
Comment (thorstenhater) [reply arbor/include/arbor/mechanism.hpp:57]: I had a discussion with @halfflat  some time ago from which I concluded that indeed the plan is to get rid of the `typedef`s  where possible.

File: arbor/backends/event_delivery.md (hunk 1 above)
Comment (thorstenhater) [reply arbor/backends/event_delivery.md:117]: Regarding the `apply_events` name, I had the compiler complain about 'overloaded virtuals' after moving the definition to `concrete_mechanism`. Although that have gone away after various iterations of clean-up.

File: arbor/backends/multicore/mechanism.cpp (hunk 3 above)
Comment (thorstenhater) [reply arbor/backends/multicore/mechanism.cpp:165]: Done

File: modcc/mechanism.hpp (hunk 5 above)
Comment (thorstenhater) [reply modcc/mechanism.hpp]: Done

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Comment by haampie** [comment]
Hm, when CI was triggered here #1365 was not merged yet and gitlab was already configured to use the new path.

bors try-

**Comment by haampie** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Comment by thorstenhater** [comment]
bors try

**Review by bcumming — APPROVED** [review]

**Comment by bcumming** [comment]
bors try

**Pull request merged**